### PR TITLE
Added error message for E0284

### PR DESCRIPTION
**Pull request by imbrem**
Work on #61137

### CODE REVIEW

File: src/librustc/error_codes.rs [hunk 1]
```diff
@@ -1207,6 +1207,52 @@ fn main() {
...
+    Ok(true)
+}
+```
+
+Here we have an iterator `results` over `Result<bool, ()>`.
+Hence, `results.collect()` can return any type implementing
+`FromIterator<Result<bool, ()>>`. On the other hand, the
+`?` operator can accept any type implementing `Try`.
+
+The user of this code probably wants `collect()` to return a
```
Comment (GuillaumeGomez) [inline src/librustc/error_codes.rs]: "user" -> "owner" maybe?

File: src/librustc/error_codes.rs [hunk 2]
```diff
@@ -1207,6 +1207,52 @@ fn main() {
...
+    let results = [Ok(true), Ok(false), Err(())].iter().cloned();
+    let v = {
+        let temp : Result<Vec<bool>, ()> = results.collect();
+        temp?
+    };
+    // Do things with v...
+    Ok(true)
+}
+```
+Note that the type of `v` can now be inferred from the type of `temp`
```
Comment (GuillaumeGomez) [inline src/librustc/error_codes.rs]: Please add an empty line before this one.

Comment (GuillaumeGomez) [reply src/librustc/error_codes.rs]: Also, please add the missing dot at the end of the sentence.

File: src/librustc/error_codes.rs [hunk 3]
```diff
@@ -1207,6 +1207,52 @@ fn main() {
...
+
+
```
Comment (GuillaumeGomez) [inline src/librustc/error_codes.rs]: Please remove those two empty lines.

**Comment by GuillaumeGomez** [comment]
Looks good to me. Could an english native double-check if that I didn't miss something please?

cc @rust-lang/docs

File: src/librustc/error_codes.rs (hunk 1 above)
Comment (imbrem) [reply src/librustc/error_codes.rs]: Author?

**Comment by imbrem** [comment]
I incorporated the suggested changes on my local fork and pushed up. Did I do it right?

File: src/librustc/error_codes.rs (hunk 2 above)
Comment (imbrem) [reply src/librustc/error_codes.rs]: Done!

File: src/librustc/error_codes.rs (hunk 3 above)
Comment (imbrem) [reply src/librustc/error_codes.rs]: Done!

**Comment by estebank** [comment]
You'll need to run `./x.py test src/test/ui --stage 1 --bless` to update the ui tests, as now that the error has an explanation `rustc` recommends reading it.

File: src/librustc/error_codes.rs [hunk 4]
```diff
@@ -1207,6 +1207,51 @@ fn main() {
...
+This error occurs when the compiler is unable to unambiguously infer the
+return type of a function or method which is generic on return type, such
+as the `collect` method for `Iterator`s.
+
+For example:
+
+```compile_fail,E0284
+fn foo() -> Result<bool, ()> {
+    let results = [Ok(true), Ok(false), Err(())].iter().cloned();
+    let v : Vec<bool> = results.collect()?;
```
Comment (aloucks) [inline src/librustc/error_codes.rs]: Usually you don't put a space between the variable and the semicolon.

```rust
let v: Vec<bool> = results.collect()?;
...
let temp: Result<Vec<bool>, ()> = results.collect();
```

Example:

https://github.com/rust-lang/rust/pull/61220/files#diff-e6786747070ada32f620caf54905df92R1117

Comment (imbrem) [reply src/librustc/error_codes.rs]: OK changed it.

**Comment by imbrem** [comment]
@estebank did I do it right?

File: src/librustc/error_codes.rs [hunk 5]
```diff
@@ -1207,6 +1207,51 @@ fn main() {
...
+`T::Ok == Vec<bool>`. Hence, this code is ambiguous and an error
+is returned.
+
+To resolve this error, use a concrete type for the intermediate expression:
+
+```
+fn foo() -> Result<bool, ()> {
+    let results = [Ok(true), Ok(false), Err(())].iter().cloned();
+    let v = {
+        let temp : Result<Vec<bool>, ()> = results.collect();
```
Comment (aloucks) [inline src/librustc/error_codes.rs]: @imbrem Missed this one

Comment (imbrem) [reply src/librustc/error_codes.rs]: Done! Sorry it took so long

**Pull request merged**